put down diceroll automatically

mattolbert opened the pull request:
-taught program to put down the diceroll automatically if that number is available
-no longer asks if user wants 1 die or 2, just automatically switches to 1 when that option arises

mattolbert committed:
put down diceroll automatically

-taught program to put down the diceroll automatically if that number is available
-no longer asks if user wants 1 die or 2, just automatically switches to 1 when that option arises

mcdaly commented on ShutTheBox.py:
```diff
@@ -25,8 +25,8 @@ def has_duplicates(seq):
   if 7 not in nums and 8 not in nums and 9 not in nums:
-    dice_choice = input("Would you like to roll 1 die or 2? ")
-    rollvalue = roll_die(dice_choice)
+    # dice_choice = input("Would you like to roll 1 die or 2? ")
```
Instead of commenting this out, how could you make the code work as you want while also, leaving the option for the user to play?

mattolbert replied:
are you saying, for example, at the beginning of each game, ask "Do you want to play or do you want the comp to play automatically?" and have both options available?

mcdaly replied:
Yes!

mattolbert replied:
Seems doable!

mcdaly commented on ShutTheBox.py:
```diff
@@ -38,26 +38,31 @@ def has_duplicates(seq):
-        any_nums_down_already = any(i not in nums for i in choicelist)
-        if any_nums_down_already:
-            print("Your number choice has already been put down")
+        choice = input("Pick number(s) to remove, separated by commas:")
```
It is still asking me for an input

mattolbert replied:
For the moment, I have it play automatically if the number the dice rolled is itself available, but if not (if that number is gone, or if it rolls 10, 11, or 12) it still asks for input. Was it behaving differently than that?

mcdaly replied:
nope. that is what is happening

mcdaly reviewed (changes requested):
A good start to automating. More work should be done in this PR before merging to get it fully automated for a game.

mattolbert committed:
Combo_List

- combo function returns list of possible combos
- program automatically puts down the 1st available combo, which is either the diceroll num itself or the highest/lowest pair of nums available

mattolbert committed:
Play_10_Times

-added a counter to run the game automatically 10 times and save the 10 scores to a list

mattolbert committed:
Added_score_wording

-added score wording too

mattolbert committed:
Both_User_and_Comp_can_play

-added option at the top for either the user to play or the comp to play
-if comp, user sets how many games to sim
-both options seem to run great I think.

mattolbert committed:
fixed_that_variable

whoops

mattolbert committed:
Added losing streak stats

-added losing streak stats to calculate longest and shortest distance between wins when simming the game

mcdaly commented on ShutTheBox.py:
```diff
@@ -1,75 +1,130 @@
 
-nums = list(range(1, 10))
-print(nums)
+player = int(input("If you'd like to play Shut the Box, press 1. If you'd like the computer to play, press 2:"))
```
Nice! Exactly what I was thinking and nice addition of asking how many times to play

mcdaly commented on ShutTheBox.py:
```diff
@@ -1,75 +1,130 @@
-    die2 = random.randint(1, 6)
-    rollval = die1 + die2
-  return rollval
+  def does_it_add_up(nums, target_num):
```
It is a bit strange to be defining function within the whole loop.
I would recommend doing one of two things
1. Move all your functions to the top of the file before your code actually starts
2. Move your functions into a different file and import the functions! (Far better for coding standards)

mcdaly commented on ShutTheBox.py:
```diff
@@ -1,75 +1,130 @@
+        #if any combinations of available nums match the dice roll, add that combo to the combos list
+        if sum(combo) == target_num:
+          combos.append(combo)
+    #if the combos list isn't empty, play on
```
I understand why you have all this code below and you return the tuple… 
But you should be able to get the same functionality in your code just by returning the ‘combo’ list.
Think about how you could do that and what check would be needed out of this function to get the equivalent of your True/False value

mattolbert replied:
can you give me a bit more detail or a nudge for this one?

mcdaly replied:
The True/False is only used in two places in the main code (and it can really be reduced to one by changing to an else statement.
Instead of passing back this extra variable, how can you use the state of the combo list in place of how the first part of the tuple is used when this is returned?

my reasoning for this is that passing back tuples is possible but it can get messy quick and cause trouble since it relies on the order of the output to be consistent when referencing it in other parts of the code. This is not a big deal here since it is only called once

mcdaly commented on ShutTheBox.py:
```diff
@@ -1,75 +1,130 @@
-  return rollval
+  def does_it_add_up(nums, target_num):
+    combos = []
+    for r in range(1, len(nums) + 1):
```
What is this variable name? ‘r’? Feel free to spend a couple extra characters describing the variable for readability

mattolbert replied:
great feedback, i'll do that going forward

mcdaly commented on ShutTheBox.py:
```diff
@@ -1,75 +1,130 @@
-      choicemath = sum(choicelist)
+  while nums:
+    if 7 not in nums and 8 not in nums and 9 not in nums:
+      if player == 1:
```
Nice!

mcdaly commented on ShutTheBox.py:
```diff
@@ -1,75 +1,130 @@
-      #Ensure the user didn't type the same number in twice
-      elif has_duplicates(choicelist) == True:
-        print("You can't put the same number down twice")
+    if player == 1:
```
Excellent only showing these steps with a single play!

mcdaly commented on ShutTheBox.py:
```diff
@@ -1,75 +1,130 @@
+      print("You won! You shut the box!")
+      
+  counter += 1
+  #my list of scores were all strings like ('123', '45'), etc. but i wanted them to be ints so I could give some of the stats below
```
Good start to the analysis

mcdaly reviewed (changes requested):
Excellent updates!! I really like the ability to play automatically for a given number of times and get some metrics.
Now that your code is starting to get long, it may be worth moving all the code into functions to help clean it up but that could be for another PR.
As always, let me know if you have any questions on my comments! Good luck!

mcdaly commented on ShutTheBox.py:
```diff
@@ -1,75 +1,130 @@
+    print(f"Your worst score was {max(score_list)} and you actually got that score {max_bad_score} times. Yikes!")
+  else:
+    print(f"Your worst score was {max(score_list)}. Yikes!")
+  print(f"Your longest stretch between wins was {max(losing_streak_list)} games and your shortest stretch was {min(losing_streak_list)}.")
```
this line breaks if you never win a game

mattolbert replied:
dang good catch

mcdaly commented on ShutTheBox.py:
```diff
@@ -1,75 +1,130 @@
+  else:
+    print(f"Your worst score was {max(score_list)}. Yikes!")
+  print(f"Your longest stretch between wins was {max(losing_streak_list)} games and your shortest stretch was {min(losing_streak_list)}.")
+  
```
another fun stat could be timing each game and outputting the average time for each game

mattolbert replied:
time in terms of, like, actual time on a clock? or in terms of number of rolls?

mcdaly replied:
I was thinking clock time, but either could be interesting

mattolbert replied:
added roll counter in newest commit, haven't played w/ time yet but i'm gonna close this for now and think about adding it later or not

mattolbert committed:
Roll Counter

-moved functions up to the top and out of the while loop
- added a roll counter to tally up the rolls per game and give an avg.
-fixed bug about 'longest stretch between wins' breaking if the comp never wins
